enclave-runner/src/usercalls/mod.rs

```diff
@@ -346,37 +385,228 @@ impl EnclaveState {
+                        }
+                        match result {
+                            Ok(ret) => {
+                                if enclave.work_sender.lock().unwrap().as_ref().unwrap()
+                                    .send(Work {tcs:state, entry: CoEntry::Resume(usercall, ret )})
+                                    .is_err() {
+                                    panic!("Work sender couldn't send data to receiver")
+                                }
+                                continue;
+                            },
+                            Err(EnclaveAbort::Exit { panic: true }) => {
+                                #[cfg(unix)]
+                                trap_attached_debugger(usercall.tcs.address().0 as _);
+                                my_result = Err(EnclaveAbort::Exit {
+                                    panic: EnclavePanic::from(buf.into_inner()),
+                                });
+                            },
+                            Err(EnclaveAbort::Exit { panic: false }) => {
+                                if state.mode != EnclaveEntry::ExecutableMain {
+                                    erased_tcs = Some(usercall.tcs);
+                                }
+                                my_result = Ok((0,0));
+                            },
+                            Err(EnclaveAbort::IndefiniteWait) => my_result = Err(EnclaveAbort::IndefiniteWait),
+                            Err(EnclaveAbort::InvalidUsercall(n)) => my_result = Err(EnclaveAbort::InvalidUsercall(n)),
+                            Err(EnclaveAbort::MainReturned) => my_result = Err(EnclaveAbort::MainReturned),
+                            Err(EnclaveAbort::Secondary) => my_result = Err(EnclaveAbort::Secondary),
+                        }
+                    },
+                    CoResult::Return((tcs,v1, v2)) => {
+                        match state.mode {
+                            EnclaveEntry::Library => {
+                                erased_tcs = Some(tcs);
+                                my_result = Ok((v1, v2));
+                            }
+                            EnclaveEntry::ExecutableMain => {
+                                main_result = Err(EnclaveAbort::MainReturned);
+                                break 'outer;
+                            }
+                            EnclaveEntry::ExecutableNonMain => {
+                                assert_eq!(
+                                    (v1, v2),
+                                    (0, 0),
+                                    "Expected enclave thread entrypoint to return zero"
+                                );
+                                erased_tcs = Some(tcs);
+                                my_result = Ok((0,0));
+                            }
+                        }
+                    }
+                };
+                if state.mode != EnclaveEntry::ExecutableNonMain && my_result.is_err() {
```

Comment:
Combine this with the match below so it's easier to see what's handled and what's not, e.g.:
```rust
match (my_result, enclave.kind, state.mode) {
    ...
}
```